h2o-core/src/main/java/water/ExternalFrameReaderClient.java

```diff
@@ -0,0 +1,167 @@
+ * }
+ * </pre>
+ */
+public class ExternalFrameReaderClient {
+
+    private AutoBuffer ab;
+    private String keyName;
+    private int chunkIdx;
+    private int[] selectedColumnIndices;
+    private ByteChannel channel;
+    private int numRows;
+    private byte[] expectedTypes = null;
+
+    public ExternalFrameReaderClient(ByteChannel channel, String keyName, int chunkIdx, byte[] expectedTypes, int[] selectedColumnIndices) throws IOException{
+        this.channel = channel;
+        this.keyName = keyName;
+        this.chunkIdx = chunkIdx;
+        this.expectedTypes = expectedTypes;
+        this.selectedColumnIndices = selectedColumnIndices;
+        this.ab = prepareAutoBuffer();
+        // once we send H2O all information it needs to prepare for reading, it sends us back number of rows
+        this.numRows = ab.getInt();
+    }
+
+    public int getNumRows(){
+        return numRows;
+    }
+
+    public boolean readBoolean(){
+        return ab.getZ();
+    }
+    public byte readByte(){
+        return ab.get1();
+    }
+    public char readChar(){
+        return ab.get2();
+    }
+    public short readShort(){
+        return ab.get2s();
+    }
+    public int readInt(){
+        return ab.getInt();
+    }
+    public long readLong(){
+        return ab.get8();
+    }
+    public float readFloat(){
+        return ab.get4f();
+    }
+    public double readDouble(){
+        return ab.get8d();
+    }
+    public String readString(){
+        return ab.getStr();
+    }
+    public Timestamp readTimeStamp(){
+        return new Timestamp(ab.get8());
+    }
+
+    public boolean isNA(boolean data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(byte data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(short data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(char data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(int data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(long data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(float data){
+        return ExternalFrameUtils.isNA(data);
+    }
+
+    public boolean isNA(double data){
+        return ExternalFrameUtils.isNA(data);
+    }
+
+    public boolean isNA(Timestamp data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    public boolean isNA(String data){
+        return ExternalFrameUtils.isNA(ab, data);
+    }
+
+    /**
+     * This method ensures the application waits for all bytes to be received before continuing in the control flow.
+     *
+     * It has to be called at the end of reading.
+     */
+    public void waitUntilAllReceived(){
+        // confirm that all has been done before proceeding with the computation
+        assert(ab.get1() == ExternalFrameHandler.CONFIRM_READING_DONE);
```

Comment:
I will separate this into 2 lines, it's better to have assert without side effects
```
// blocking call
byte controlByte = ab.get1() 
assert(controlByte== ExternalFrameHandler.CONFIRM_READING_DONE);
```